portfolio/src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -0,0 +1,128 @@
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles user accounts */
+@WebServlet("/user")
+public class UserServlet extends HttpServlet {
+
+  /**
+   * Handles the GET requests to "/user" path.
+   * Returns a JSON object with the user information.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    User user;
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      String userId = userService.getCurrentUser().getUserId();
+      String userEmail = userService.getCurrentUser().getEmail();
+      String urlToRedirectToAfterUserLogsOut = "/";
+      String logoutURL = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+
+      // Load user data from the database
+      Entity entity = getUser(userId);
+      // If the user data has not been stored in the database, then create the corresponding entity
+      if (entity == null) {
+        entity = new Entity("User", userId);
+        entity.setProperty("id", userId);
+        entity.setProperty("email", userEmail);
+        // By default, set the username to the email address without the domain 
+        entity.setProperty("username", (userEmail.split("@")[0]));
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(entity);
+      }
+    
+      user = new AuthenticatedUser(userId, 
+                                   userEmail, 
+                                   (String) entity.getProperty("username"), 
+                                   logoutURL);
+    } else {
+      String urlToRedirectToAfterUserLogsIn = "/#comments";
+      String loginURL = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+
+      user = new AnonymousUser(loginURL);
+    }
+
+    // Convert the user to JSON
+    Gson gson = new Gson();
+    String json = gson.toJson(user);
+
+    // Send the JSON as the response
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Returns the User entity from the database based on the user's ID
+   */
+  public Entity getUser(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("User")
+            .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+
+    return entity;
+  }
+
+  /**
+   * Handles the POST requests to "/user" path.
+   * Changes the username in the database with the newly submitted username.
+   */
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    // Make sure the user is authenticated
+    if (userService.isUserLoggedIn()) {
```

Comment:
What should happen if the user hasn't logged in? Would it be better to respond some messages?